llvm/lib/CAS/CASOptions.cpp

```diff
@@ -92,3 +93,59 @@ void CASOptions::ensurePersistentCAS() {
     break;
   }
 }
+
+using llvm::yaml::IO;
+using llvm::yaml::MappingTraits;
+
+struct CASIDFileImpl {
+  std::string CASPath;
+  std::string CASID;
+};
+
+namespace llvm {
+namespace yaml {
+
+template <> struct MappingTraits<CASIDFileImpl> {
+  static void mapping(IO &IO, CASIDFileImpl &Value) {
+    IO.mapTag("!cas-file-v1", true);
+    IO.mapRequired("cas-id", Value.CASID);
+    IO.mapRequired("cas-path", Value.CASPath);
+  }
+};
+
+} // namespace yaml
+} // namespace llvm
+
+void CASOptions::writeCASIDFile(raw_ostream &OS, const CASID &ID) const {
+  CASIDFileImpl Output;
+  raw_string_ostream SS(Output.CASID);
+  ID.print(SS);
+  // Use the CASPath from the cache as the frozen CASOption have the path wiped.
+  Output.CASPath = Cache.Config.CASPath;
```

Comment:
This still feels like an information leak to me. Now anyone that wants to can sniff out the CAS configuration by writing to and reading from a `SmallVector` / `raw_svector_ostream`, and there's no layering to prevent this output file from being cached by something that wants to be independent of the CAS location.

Instead, this should use `CASPath`, with a comment "not using Cache.Config.CASPath, since that would leak the hidden configuration", or take an optional `CASConfiguration*=nullptr` as an argument to pass in settings. Th elatter, I'm thinking:
```
void CASOptions::writeCASIDFile(raw_ostream &OS, const CASID &ID, const CASConfiguration *ConfigToSerialize) const {
  if (!ConfigToSerialize)
    ConfigToSerialize = this;
```

For clang, you can just save this up a level, and pass it in as a parameter. Here's the current code. Note the TODO:
```
  // Hide the CAS configuration, canonicalizing it to keep the path to the
  // CAS from leaking to the compile job, where it might affecting its
  // output (e.g., in a diagnostic).
  //
  // TODO: Extract CASOptions.Path first if we need it later since it'll
  // disappear here.
  CAS = Invocation.getCASOpts().getOrCreateCASAndHideConfig(Diags);
```
You can replace it with:
```
  CASOptionsConfig = Invocation.getCASOpts();
  CAS = Invocation.getCASOpts().getOrCreateCASAndHideConfig(Diags);
```
adding `CASConfiguration` field to CompileJobCache, and then pass in the saved config as an extra argument when writing the file, or:
```
  CAS = Invocation.getCASOpts().getOrCreateCAS(Diags);
  CASOpts = Invocation.getCASOpts();
  Invocation.getCASOpts().hideConfig(); // new function to implement.
```
adding `CASOptions` field to CompileJobCache.

Reply:
I wonder, maybe only `CASConfiguration` should be sunk down into LLVM, and this could just be a free function that takes `const CASConfiguration&` by argument. Then there's no need to override it.
- Clang can save the config in the CompileJobCache and pass it in when writing out the file in `CompileJobCache::finishComputedResult()`
- `CASOptions` is free to reference clang-specific things like `DiagnosticsEngine`

Reply:
Technically, I didn't change CASOption API at all so what is leaking is what is currently leaking. The interface needed to write CASID file needs to leak the path as the target audience is not clang but binary tools that are not running inside a build system like `nm`. Here my current thinking. 

* Sink CASConfiguration into llvm, and it is ok to be leaked.
* CASOption is clang specific. It is not inheriting from CASConfiguration, but use that as a private field, so CASOption is created from CASConfiguration but nothing can get the CASPath out of the CASOption. If clang wants to write a CASID file, the CacheJobAction needs to keep track of the CASConfiguration used to create CASOption and use that.

Reply:
> Technically, I didn't change CASOption API at all so what is leaking is what is currently leaking.

I disagree. Technically, you changed the CASOptions API and started leaking the hidden configuration. Previously if someone received `CASOptions`, they could not accidentally depend on a hidden config. Now they can! E.g.:
```
auto doSomething(const CASOptions &HiddenOpts, OtherArgs) {
  CASID Result1 = compute1(HiddenOpts, OtherArgs); // Is Result1 tainted? Hopefully not...
  SmallVector<char> Result2;
  raw_svector_ostream OS(Result2);
  HiddenOpts.writeCASIDFile(OS, Result1); // Result2 has been tainted (accidental leak)
  auto Result3 = compute3(HiddenOpts, Result2); // Result3 has been tainted (accidental leak)

  // Direct leak: read the config back. Depending on how deeply this logic is buried,
  // this could also be an accidental leak.
  CASOptions FoundOpts;
  FoundOpts.createFromCASIDFile(MemoryBufferRef(Result2));

  // ...
```

> The interface needed to write CASID file needs to leak the path as the target audience is not clang but binary tools that are not running inside a build system like nm.

I understand that there is a practical use case with a target audience that does not include clang caching. The point of contention is on how to architect it such that it doesn't have accidental side effects on clang caching.

I think it's important that the architecture for caching is hermetic (leak-free). It needs to be clear and obvious what the inputs are. If there's something we don't want to affect results, then we cannot let it be observable, or we cannot trust that the computations are pure. If we can't trust the computations to be pure, it's not sound to cache them.

> Here my current thinking. [snip]

Yup, I think that matches my thinking. To be clear, if `CASOptions` has hidden its configuration, then there should be no way to observe that configuration (and have it become an accidental input). But passing the original (saved) `CASConfiguration` *as an input* to some other computation (such a function that creates a file from a CASID+CASConfig) sounds fine. It'll affect that output, but since the config is an explicit input to the computation, that seems fine.

However, note that anything checking for reproducible builds will see that the environment (path to the CAS) has affected the object file. There's a tradeoff there.